<h1><center>Laboratorio 6: La solicitud de Sergio 🤗</center></h1>

<center><strong>MDS7202: Laboratorio de Programación Científica para Ciencia de Datos</strong></center>

### Cuerpo Docente:

- Profesores: Ignacio Meza, Sebastián Tinoco
- Auxiliar: Catherine Benavides
- Ayudante: Nicolás Ojeda, Eduardo Moya

### Equipo: SUPER IMPORTANTE - notebooks sin nombre no serán revisados

- Nombre de alumno 1: Diego Cornejo


### **Link de repositorio de GitHub:** `https://github.com/diegonachon/mds7202/tree/main/lab6`

## Temas a tratar
- Aplicar Pandas para obtener características de un DataFrame.
- Aplicar Pipelines y Column Transformers.
- Utilizar diferentes algoritmos de cluster y ver el desempeño.

## Reglas:

- **Grupos de 2 personas**
- Cualquier duda fuera del horario de clases al foro. Mensajes al equipo docente serán respondidos por este medio.
- Prohibidas las copias.
- Pueden usar cualquer matrial del curso que estimen conveniente.
- Código que no se pueda ejecutar, no será revisado.

### Objetivos principales del laboratorio
- Comprender cómo aplicar pipelines de Scikit-Learn para generar clusters.
- Familiarizarse con plotly.

El laboratorio deberá ser desarrollado sin el uso indiscriminado de iteradores nativos de python (aka "for", "while"). La idea es que aprendan a exprimir al máximo las funciones optimizadas que nos entrega `numpy`, las cuales vale mencionar, son bastante más eficientes que los iteradores nativos sobre arreglos (*o tensores*).

## Descripción del laboratorio

<center>
<img src="https://i.pinimg.com/originals/5a/a6/af/5aa6afde8490da403a21601adf7a7240.gif" width=400 />

En el corazón de las operaciones de Aerolínea Lucero, Sergio, el gerente de análisis de datos, reunió a un talentoso equipo de jóvenes científicos de datos para un desafío crucial: segmentar la base de datos de los clientes. “Nuestro objetivo es descubrir patrones en el comportamiento de los pasajeros que nos permitan personalizar servicios y optimizar nuestras campañas de marketing,” explicó Sergio, mientras desplegaba un amplio rango de datos que incluían desde hábitos de compra hasta opiniones sobre los vuelos.

Sergio encargó a los científicos de datos la tarea de aplicar técnicas avanzadas de clustering para identificar distintos segmentos de clientes, como los viajeros frecuentes y aquellos que eligen la aerolínea para celebrar ocasiones especiales. La meta principal era entender profundamente cómo estos grupos perciben la calidad y satisfacción de los servicios ofrecidos por la aerolínea.

A través de un enfoque meticuloso y colaborativo, los científicos de datos se abocaron a la tarea, buscando transformar los datos brutos en valiosos insights que permitirían a Aerolínea Lucero no solo mejorar su servicio, sino también fortalecer las relaciones con sus clientes mediante una oferta más personalizada y efectiva.

## Importamos librerias utiles 😸

In [1]:
import numpy as np
import pandas as pd

from sklearn import datasets

import plotly.express as px
import plotly.graph_objects as go
from plotly.subplots import make_subplots

## 1. Estudio de Performance 📈



<center>
<img src="https://user-images.githubusercontent.com/57133330/188281408-c67df9ee-fd1f-4b37-833b-f02848f1ce02.gif" width=300>

Don Sergio les ha encomendado su primera tarea: analizar diversas técnicas de clustering. Su objetivo es entender detalladamente cómo funcionan estos métodos en términos de segmentación y eficiencia en tiempo de ejecución.

Analice y compare el desempeño, tiempo de ejecución y visualizaciones de cuatro algoritmos de clustering (k-means, DBSCAN, Ward y GMM) aplicados a tres conjuntos de datos, incrementando progresivamente su tamaño. Utilice Plotly para las gráficas y discuta los resultados tanto cualitativa como cuantitativamente.

Uno de los requisitos establecidos por Sergio es que el análisis se lleve a cabo utilizando Plotly; de no ser así, se considerará incorrecto. Para facilitar este proceso, se ha proporcionado un código de Plotly que puede servir como base para realizar las gráficas. Apóyese en el código entregado para efectuar el análisis y tome como referencia la siguiente imagen para realizar los gráficos:

<img src='https://gitlab.com/imezadelajara/datos_clase_7_mds7202/-/raw/main/misc_images/Screenshot_2024-04-26_at_9.10.44_AM.png' width=900 />

En el gráfico se visualizan en dos dimensiones los diferentes tipos de datos proporcionados en `datasets`. Cada columna corresponde a un modelo de clustering diferente, mientras que cada fila representa un conjunto de datos distinto. Cada uno de los gráficos incluye el tiempo en segundos que tarda el análisis y la métrica Silhouette obtenida.

Para ser más específicos, usted debe cumplir los siguientes objetivos:
1. Generar una función que permita replicar el gráfico expuesto en la imagen (no importa que los colores calcen). [4 puntos]
2. Ejecuta la función para un `n_samples` igual a 1000, 5000, 20000. [2 puntos]
3. Analice y compare el desempeño, tiempo de ejecución y visualizaciones de cuatro algoritmos de clustering. [4 puntos]


> ❗ Tiene libertad absoluta de escoger los hiper parámetros de los cluster, sin embargo, se recomienda verificar el dominio de las variables para realizar la segmentación.
> ❗ Recuerde que es obligatorio el uso de plotly.


**<font color="red">DISCLAIMER!</font>**

**<font color="red">POR TEMAS DE ALMACENAMIENTO, SE COMENTARON LOS GRÁFICOS GENERADOS</font>**

**<font color="red">FIN DE DISCLAIMER!</font>**

In [2]:
"""
En la siguiente celda se crean los datos ficticios a usar en la sección 1 del lab.
❗No realice cambios a esta celda
"""

# Datos a utilizar

# Configuracion
n_samples = 5000

# Lunas
moons = datasets.make_moons(n_samples=n_samples, noise=0.05, random_state=30)

# Blobs
blobs = datasets.make_blobs(n_samples=n_samples, random_state=172)

# Datos desiguales
transformation = [[0.6, -0.6], [-0.4, 0.8]]
mutated = (np.dot(blobs[0], transformation), blobs[1])

# Generamos Dataset
datasets = {
    'moons':{
        'x': moons[0], 'classes': moons[1], 'n_cluster': 2
    },
    'blobs':{
        'x': blobs[0], 'classes': blobs[1], 'n_cluster': 3
    },
    'mutated':{
        'x': mutated[0], 'classes': mutated[1], 'n_cluster': 3
    }
}

In [3]:
# # Función scatter
# def plot_scatter(x, y, color):
#     fig = go.Figure(data=[go.Scatter(
#         x=x, y=y, mode='markers',
#         marker=dict(
#             size=8,
#             color=color,  # corresponde a una variable por el cual colorear
#             showscale=False  # esconde la escala de color
#         )
#     )])
#     return fig

# # Multiples plots con plotly
# fig = make_subplots(rows=3, cols=1)
# fig.add_trace(
#     plot_scatter(
#         x=datasets['moons']['x'][:, 0],
#         y=datasets['moons']['x'][:, 1],
#         color=datasets['moons']['classes']
#     )._data[0],
#     row=1, col=1
# )
# fig.add_trace(
#     plot_scatter(
#         x=datasets['blobs']['x'][:, 0],
#         y=datasets['blobs']['x'][:, 1],
#         color=datasets['blobs']['classes']
#     )._data[0],
#     row=2, col=1
# )
# fig.add_trace(
#     plot_scatter(
#         x=datasets['mutated']['x'][:, 0],
#         y=datasets['mutated']['x'][:, 1],
#         color=datasets['mutated']['classes']
#     )._data[0],
#     row=3, col=1
# )
# fig.update_layout(
#     showlegend=False, template='simple_white'
# )

**Respuesta**

In [3]:
def plot_scatter(x, y, color):
    fig = go.Figure(data=[go.Scatter(
        x=x, y=y, mode='markers',
        marker=dict(
            size=8,
            color=color,  # corresponde a una variable por el cual colorear
            showscale=False  # esconde la escala de color
        )
    )])
    return fig

In [4]:
from sklearn.cluster import KMeans
from sklearn.mixture import GaussianMixture
from sklearn.cluster import AgglomerativeClustering
from sklearn.cluster import DBSCAN
from sklearn.metrics import silhouette_score
import time

In [5]:
def generate_plots(n_samples):
    from sklearn import datasets
    # Datos a utilizar
    # Configuracion
    n_samples = n_samples
    # Lunas
    moons = datasets.make_moons(n_samples=n_samples, noise=0.05, random_state=30)
    # Blobs
    blobs = datasets.make_blobs(n_samples=n_samples, random_state=172)
    # Datos desiguales
    transformation = [[0.6, -0.6], [-0.4, 0.8]]
    mutated = (np.dot(blobs[0], transformation), blobs[1])
    # Generamos Dataset
    datasets = {'moons':{'x': moons[0], 'classes': moons[1], 'n_cluster': 2},
                'blobs':{'x': blobs[0], 'classes': blobs[1], 'n_cluster': 3},
                'mutated':{'x': mutated[0], 'classes': mutated[1], 'n_cluster': 3}}
    
    X_moons = moons[0]
    X_blobs = blobs[0]
    X_mutat = mutated[0]
    
    silhouettes = np.zeros(12)
    times = np.zeros(12)
    
    dict_algorithms = ['kmeans', 'gmm', 'ward', 'dbscan']
    
    # Multiples plots con plotly
    fig = make_subplots(rows=3, 
                        cols=4, 
                        subplot_titles = ('a', ) * 12,
                        column_titles = ['KMeans', 'GMM', 'WARD', 'DBSCAN'])
    
    for algorithm in dict_algorithms:
        
        if algorithm == 'kmeans':
            list_indexs = 0
            row = 1
            for data in datasets:
                time_start = time.time()
                kmean = KMeans(n_clusters = datasets[data]['n_cluster'], random_state = 666)
                kmean.fit(datasets[data]['x'])
                kmean_labels = kmean.labels_
                kmean_silhoutte = silhouette_score(datasets[data]['x'], kmean_labels)
                time_end = time.time() - time_start
                silhouettes[list_indexs] = kmean_silhoutte
                times[list_indexs] = time_end
                
                fig.add_trace(
                plot_scatter(
                    x = datasets[data]['x'][:, 0],
                    y = datasets[data]['x'][:, 1],
                    color = kmean_labels
                )._data[0],
                row = row, col = 1
                )
                list_indexs += 4
                row += 1
                                
        elif algorithm == 'gmm':
            list_indexs = 1
            row = 1
            for data in datasets:
                time_start = time.time()
                gmm = GaussianMixture(n_components = datasets[data]['n_cluster'], random_state = 666)
                gmm.fit(datasets[data]['x'])
                gmm_labels = gmm.predict(datasets[data]['x'])
                gmm_silhoutte = silhouette_score(datasets[data]['x'], gmm_labels)
                time_end = time.time() - time_start
                silhouettes[list_indexs] = gmm_silhoutte
                times[list_indexs] = time_end
                
                
                fig.add_trace(
                plot_scatter(
                    x = datasets[data]['x'][:, 0],
                    y = datasets[data]['x'][:, 1],
                    color = gmm_labels
                )._data[0],
                row = row, col = 2
                )
                list_indexs += 4
                row += 1
                
        elif algorithm == 'ward':
            list_indexs = 2
            row = 1
            for data in datasets:
                time_start = time.time()
                ward = AgglomerativeClustering(n_clusters = datasets[data]['n_cluster'], linkage='ward')
                ward.fit(datasets[data]['x'])
                ward_labels = ward.labels_
                ward_silhoutte = silhouette_score(datasets[data]['x'], ward_labels)
                time_end = time.time() - time_start
                silhouettes[list_indexs] = ward_silhoutte
                times[list_indexs] = time_end
                fig.add_trace(
                plot_scatter(
                    x = datasets[data]['x'][:, 0],
                    y = datasets[data]['x'][:, 1],
                    color = ward_labels
                )._data[0],
                row = row, col = 3
                )
                list_indexs += 4
                row += 1
                
        else:
            list_indexs = 3
            row = 1
            min_samples = [3, 10, 10]
            for data in datasets:
                time_start = time.time()
                dbscan = DBSCAN(eps = 0.2, min_samples = min_samples[row - 1])
                dbscan_labels = dbscan.fit_predict(datasets[data]['x'])
                dbscan_silhoutte = silhouette_score(datasets[data]['x'], dbscan_labels)
                time_end = time.time() - time_start
                silhouettes[list_indexs] = dbscan_silhoutte
                times[list_indexs] = time_end
                fig.add_trace(
                plot_scatter(
                    x = datasets[data]['x'][:, 0],
                    y = datasets[data]['x'][:, 1],
                    color = dbscan_labels
                )._data[0],
                row = row, col = 4
                )
                list_indexs += 4
                row += 1

    for i in range(len(times)):
        title = f'{times[i]:.2f} [s] | s: {silhouettes[i]:.2f}'
        fig.layout['annotations'][i].update(text = title)
        fig.layout['annotations'][i]['font'].update(size = 8)
    
    for i in range(1,5):
        fig.layout['annotations'][-i].update(y = 1.05)     
    
    fig.update_layout(
        showlegend=False, 
        template='simple_white',
        height = 800,
        width = 1000
    )   
    
#     print('silhouettes')
#     print(silhouettes[:4])
#     print(silhouettes[4:8])
#     print(silhouettes[8:])
    
#     print('tiempos')
#     print(times[:4])
#     print(times[4:8])
#     print(times[8:])
    
    return(fig.show())

        

In [ ]:
generate_plots(1000)

In [8]:
generate_plots(5000)

In [9]:
generate_plots(20000)

**Respuesta**

Analice y compare el desempeño, tiempo de ejecución y visualizaciones de cuatro algoritmos de clustering



* Análisis por desempeño: A partir del coeficiente de Silueta, se tiene que en el algoritmo de K-means se comporta de mejor manera en los sets de moons y mutated, con cualquier cantidad de samples; mientras que en el set de blob, a niveles bajo de samples se comportan igual K-means, GMM y Ward; y a niveles alto solo K-means y GMM. Además, DBSCAN es el que obtuvo peores resultados en todos los niveles, muy probablemente por los hiperparámetros seleccionados.
    
    
* Análisis por tiempo: Se tiene que DBSCAN es el algoritmo con menor tiempo de ejecución en promedio, tanto por el tipo de datos, como por la cantidad de muestras. En cuanto al que más se demora, se tiene que a bajas cantidades de muestras, K-means en promedio se demora más, sin embargo, con alta cantidad de datos Wards es el más lento. Finalmente, entre 1.000, 5.000 y 20.000, existe un incremento de tiempo de ejecución de casi un 1.000% por muestra en cada uno de los algoritmos, a excepción de Wards, que aumenta alrededor de 2.000% entre las 5.000 y 20.000 muestras.
    
* Análisis de visualización: A medida que aumenta la cantidad de datos, se observa un engrosamiento en los clusters visibles. En cuanto a la clasificación de estos, con 2.000 y 5.000 muestras no se observan diferencias notorias en los algoritmos, a excepción de DBSCAN, donde en los datos blob y mutated se empiezan identificar las clases de manera más visual. Cuando hay 20.000 casos, en DBSCAN, con los datos de mutated, hay dos clusters pasan a ser identificados como uno, dado que la cantidad de datos aumentó e hizo que estos grupos se tocaran.

## 2. Análisis de Satisfacción de Vuelos

<center>
<img src="https://i.gifer.com/2Hci.gif" width=400 />

Habiendo entendido cómo funcionan los modelos de aprendizaje no supervisado, *Don Sergio* le encomienda estudiar la satisfacción de pasajeros al haber tomado un vuelo en alguna de sus aerolineas. Para esto, el magnate le dispone del dataset `aerolineas_licer.parquet`, el cual contiene el grado de satisfacción de los clientes frente a diferentes aspectos del vuelo. Las características del vuelo se definen a continuación:

- *Gender*: Género de los pasajeros (Femenino, Masculino)
- *Customer Type*: Tipo de cliente (Cliente habitual, cliente no habitual)
- *Age*: Edad actual de los pasajeros
- *Type of Travel*: Propósito del vuelo de los pasajeros (Viaje personal, Viaje de negocios)
- *Class*: Clase de viaje en el avión de los pasajeros (Business, Eco, Eco Plus)
- *Flight distance*: Distancia del vuelo de este viaje
- *Inflight wifi service*: Nivel de satisfacción del servicio de wifi durante el vuelo (0:No Aplicable; 1-5)
- *Departure/Arrival time convenient*: Nivel de satisfacción con la conveniencia del horario de salida/llegada
- *Ease of Online booking*: Nivel de satisfacción con la facilidad de reserva en línea
- *Gate location*: Nivel de satisfacción con la ubicación de la puerta
- *Food and drink*: Nivel de satisfacción con la comida y la bebida
- *Online boarding*: Nivel de satisfacción con el embarque en línea
- *Seat comfort*: Nivel de satisfacción con la comodidad del asiento
- *Inflight entertainment*: Nivel de satisfacción con el entretenimiento durante el vuelo
- *On-board service*: Nivel de satisfacción con el servicio a bordo
- *Leg room service*: Nivel de satisfacción con el espacio para las piernas
- *Baggage handling*: Nivel de satisfacción con el manejo del equipaje
- *Check-in service*: Nivel de satisfacción con el servicio de check-in
- *Inflight service*: Nivel de satisfacción con el servicio durante el vuelo
- *Cleanliness*: Nivel de satisfacción con la limpieza
- *Departure Delay in Minutes*: Minutos de retraso en la salida
- *Arrival Delay in Minutes*: Minutos de retraso en la llegada

En consideración de lo anterior, realice las siguientes tareas:

0. Ingeste el dataset a su ambiente de trabajo.

1. Seleccione **sólo las variables numéricas del dataset**.  Explique qué éfectos podría usar variables categóricas en un algoritmo no supervisado. Los datos se encuentran en el siguiente [enlace](https://gitlab.com/imezadelajara/datos_clase_7_mds7202/-/raw/main/aerolineas_lucer.parquet). [2 puntos]

2. Realice una visualización de la distribución de cada variable y analice cada una de estas distribuciones. [2 puntos]

3. Basándose en los gráficos, evalúe la necesidad de escalar los datos y explique el motivo de su decisión. [2 puntos]

4. Examine la correlación entre las variables mediante un correlograma. [2 puntos]

5. De acuerdo con los resultados obtenidos en 5, reduzca la dimensionalidad del conjunto de datos a cuatro variables, justificando su elección respecto a las variables que decide eliminar. [2 puntos]

**Respuesta:**

**P0**

In [7]:
df = pd.read_parquet('aerolineas_lucer.parquet')

**P1**

Usar variables categóricas en un algoritmo no supervisado puede traer distintos tipos de problemas, como por ejemplo:

- Errores en las distancias: Como estos algoritmos, por lo general, se basan en distancias para agrupar datos, si no se manejan adecuadamente las variables categóricas pueden ser malinterpretadas estas distancias, generando errores en cuanto a las agrupaciones de datos que se realizan o teniendo impacto en algún tipo de reducción de dimensionalidad que se quiera realizar, como un PCA.

- Problemas de escala: Existen algoritmos que son sensibles a la escala de las variables, por lo que si no se tratan correctamente estas variables, se pueden distorsionar los resultados obtenidos.

- Pérdida de información: Al representar las variables categóricas como números, se pierde la información de la categoría original, como pueden ser las escalas nominales.

In [8]:
df_numerics_var = df.select_dtypes(include=['number'])
df_numerics_var.head()

id  Age  Flight Distance  Inflight wifi service  \
0   70172   13              460                      3   
1    5047   25              235                      3   
2  110028   26             1142                      2   
3   24026   25              562                      2   
4  119299   61              214                      3   

   Departure/Arrival time convenient  Ease of Online booking  Gate location  \
0                                  4                       3              1   
1                                  2                       3              3   
2                                  2                       2              2   
3                                  5                       5              5   
4                                  3                       3              3   

   Food and drink  Online boarding  Seat comfort  Inflight entertainment  \
0               5                3             5                       5   
1               1                3             1                       1   
2               5                5             5                       5   
3               2                2             2                       2   
4               4                5             5                       3   

   On-board service  Leg room service  Baggage handling  Checkin service  \
0                 4               3.0                 4              4.0   
1                 1               5.0                 3              1.0   
2                 4               3.0                 4              4.0   
3                 2               5.0                 3              1.0   
4                 3               4.0                 4              3.0   

   Inflight service  Cleanliness  Departure Delay in Minutes  \
0               5.0            5                          25   
1               4.0            1                           1   
2               4.0            5                           0   
3               4.0            2                          11   
4               3.0            3                           0   

   Arrival Delay in Minutes  
0                      18.0  
1                       6.0  
2                       0.0  
3                       9.0  
4                       0.0

**P2**

In [12]:
# for col in df_numerics_var.columns:
#     print(f'''
#     fig = px.histogram(df_numerics_var, x = '{col}', title = '{col} Histogram')
#     fig.show()
#     ''')

In [9]:
df_numerics_var.describe()

id            Age  Flight Distance  Inflight wifi service  \
count  236114.000000  236114.000000    236114.000000          236114.000000   
mean    64976.757820      39.385297      1190.646891               2.729199   
std     37475.512146      15.125768       999.005937               1.326889   
min         1.000000       7.000000        31.000000               0.000000   
25%     32573.000000      27.000000       414.000000               2.000000   
50%     64945.000000      40.000000       843.000000               3.000000   
75%     97414.000000      51.000000      1744.000000               4.000000   
max    129880.000000      85.000000      4983.000000               5.000000   

       Departure/Arrival time convenient  Ease of Online booking  \
count                      236114.000000           236114.000000   
mean                            3.060598                2.758464   
std                             1.524643                1.399083   
min                             0.000000                0.000000   
25%                             2.000000                2.000000   
50%                             3.000000                3.000000   
75%                             4.000000                4.000000   
max                             5.000000                5.000000   

       Gate location  Food and drink  Online boarding   Seat comfort  \
count  236114.000000   236114.000000    236114.000000  236114.000000   
mean        2.978536        3.202478         3.250218       3.439584   
std         1.278132        1.328406         1.349914       1.317642   
min         0.000000        0.000000         0.000000       0.000000   
25%         2.000000        2.000000         2.000000       2.000000   
50%         3.000000        3.000000         3.000000       4.000000   
75%         4.000000        4.000000         4.000000       5.000000   
max         5.000000        5.000000         5.000000       5.000000   

       Inflight entertainment  On-board service  Leg room service  \
count           236114.000000     236114.000000     236114.000000   
mean                 3.360131          3.383416          3.353338   
std                  1.332025          1.287594          1.315401   
min                  0.000000          0.000000          0.000000   
25%                  2.000000          2.000000          2.000000   
50%                  4.000000          4.000000          4.000000   
75%                  4.000000          4.000000          4.000000   
max                  5.000000          5.000000          5.000000   

       Baggage handling  Checkin service  Inflight service    Cleanliness  \
count     236114.000000    236114.000000     236114.000000  236114.000000   
mean           3.632991         3.303980          3.641978       3.286396   
std            1.180697         1.266034          1.174758       1.311961   
min            1.000000         0.000000          0.000000       0.000000   
25%            3.000000         3.000000          3.000000       2.000000   
50%            4.000000         3.000000          4.000000       3.000000   
75%            5.000000         4.000000          5.000000       4.000000   
max            5.000000         5.000000          5.000000       5.000000   

       Departure Delay in Minutes  Arrival Delay in Minutes  
count               236114.000000             236114.000000  
mean                    14.765512                 15.213676  
std                     38.114330                 38.705523  
min                      0.000000                  0.000000  
25%                      0.000000                  0.000000  
50%                      0.000000                  0.000000  
75%                     12.000000                 13.000000  
max                   1592.000000               1584.000000

- **Age** : Presenta una distribución normal, donde los valores más comúnes se encuentran en el centro. Se tiene como promedio 39 años con una desviación estandar de +- 15 años; eñ 50% de los pasajeros tienen al menos 40 años, mientras que el 75% de estos tienen al menos 51 años.


In [14]:
fig = px.histogram(df_numerics_var, x = 'Age', title = 'Age Histogram')
fig.show()

* Fligh Distance: La distancia promedio de viaje es de 1.190 (no se especifica si son millas o kilometros), con una desviación estandar de +- cercana a 1.000. El 50% de los vuelos recorren al menos 843, mientras que el 75% recorren al menos 1.744. Presenta una distribución normal desplazada.

In [15]:
fig = px.histogram(df_numerics_var, x = 'Flight Distance', title = 'Flight Distance Histogram')
fig.show()

* *Inflight wifi service*: Corresponde a una variable categórica que va de 0 a 5. Presenta una distribución normal, teniendo que los niveles 2 y 3 son las más repetidas; mientras que el nivel 0 con menos.

In [16]:
fig = px.histogram(df_numerics_var, x = 'Inflight wifi service', title = 'Inflight wifi service Histogram')
fig.show()

* *Departure/Arrival time convenient*: Corresponde a una variable categórica  que va de 0 a 5. Presenta una distribución normal, teniendo el nivel 4 con mayor cantidad de observaciones; y el nivel 0 con menos.

In [17]:
fig = px.histogram(df_numerics_var, x = 'Departure/Arrival time convenient', title = 'Departure/Arrival time convenient Histogram')
fig.show()

* *Ease of Online booking*: Corresponde a una variable categórica que va de 0 a 5. Presenta una distribución normal, donde los niveles 2 y 3 son los que poseen mayor cantidad de observaciones, teniendo el 3 ligeramente más; y el 0 con menos.

In [18]:
fig = px.histogram(df_numerics_var, x = 'Ease of Online booking', title = 'Ease of Online booking Histogram')
fig.show()

* *Gate location*: Corresponde a una variable categórica que va de 0 a 5. Presenta una distribución normal, donde el nivel 3 es el que posee mayor cantidad de observaciones. Así mismo, 0 es el que menos posee, teniendo solo 2 datos, lo cuál puede indicar que son outliers o casos muy específicos.

In [19]:
fig = px.histogram(df_numerics_var, x = 'Gate location', title = 'Gate location Histogram')
fig.show()

* *Food and drink*: Corresponde a una variable categórica que va de 0 a 5. Presenta una distribución normal, teniendo una cantidad de observaciones similares los niveles 2, 3, 4 y 5, siendo 4 el que posee ligeramente más. Por otro lado, el que menos posee es 0, con 242 observaciones.

In [20]:
fig = px.histogram(df_numerics_var, x = 'Food and drink', title = 'Food and drink Histogram')
fig.show()

* *Online boarding*: Corresponde a una variable categórica que va de 0 a 5. Presenta una distribución normal, donde el nivel 4 es el que presenta mayor cantidad de observaciones; mientras que el nivel 0 posee la menor.

In [21]:
fig = px.histogram(df_numerics_var, x = 'Online boarding', title = 'Online boarding Histogram')
fig.show()

* *Seat comfort*: Corresponde a una variable categórica que va de 0 a 5. Presenta una distribución normal, donde el nivel 4 es el que presenta mayor cantidad de observaciones; mientras que el nivel 0 posee la menor. Así mismo, este último presenta solo 2 observaciones, lo que puede indicar que son outliers o casos específicos, teniendo como según menor nivel el 1.

In [22]:
fig = px.histogram(df_numerics_var, x = 'Seat comfort', title = 'Seat comfort Histogram')
fig.show()

* *Inflight entertainment*: Corresponde a una variable categórica que va de 0 a 5. Presenta una distribución normal, donde 4 es el que presenta mayor cantidad de observaciones; mientras que el nivel 0 posee la menor.

In [23]:
fig = px.histogram(df_numerics_var, x = 'Inflight entertainment', title = 'Inflight entertainment Histogram')
fig.show()

* *On-board service*: Corresponde a una variable categórica que va de 0 a 5. Presenta una distribución normal, donde 4 es el que presenta mayor cantidad de observaciones; mientras que el nivel 0 posee la menor. 

In [24]:
fig = px.histogram(df_numerics_var, x = 'On-board service', title = 'On-board service Histogram')
fig.show()

* *Leg room service*: Corresponde a una variable categórica que va de 0 a 5. Presenta una distribución normal, donde 4 es el que presenta mayor cantidad de observaciones; mientras que el nivel 0 posee la menor. 

In [25]:
fig = px.histogram(df_numerics_var, x = 'Leg room service', title = 'Leg room service Histogram')
fig.show()

* *Baggage handling*: Corresponde a una variable categórica que va de 0 a 5. Presenta una distribución normal, donde 4 es el que presenta mayor cantidad de observaciones; mientras que el nivel 0 posee la menor. 

In [26]:
fig = px.histogram(df_numerics_var, x = 'Baggage handling', title = 'Baggage handling Histogram')
fig.show()

* *Checkin service*: Corresponde a una variable categórica que va de 0 a 5. Presenta una distribución normal, donde 4 es el que presenta mayor cantidad de observaciones; mientras que el nivel 0 posee la menor. 

In [27]:
fig = px.histogram(df_numerics_var, x = 'Checkin service', title = 'Checkin service Histogram')
fig.show()

* *Inflight service*: Corresponde a una variable categórica que va de 0 a 5. Presenta una distribución normal, donde 4 es el que presenta mayor cantidad de observaciones; mientras que el nivel 0 posee la menor. 

In [28]:
fig = px.histogram(df_numerics_var, x = 'Inflight service', title = 'Inflight service Histogram')
fig.show()

* *Cleanliness*: Corresponde a una variable categórica que va de 0 a 5. Presenta una distribución normal, donde 4 es el que presenta mayor cantidad de observaciones, seguido por el 3; mientras que el nivel 0 posee la menor. 

In [29]:
fig = px.histogram(df_numerics_var, x = 'Cleanliness', title = 'Cleanliness Histogram')
fig.show()

* *Departure Delay in Minutes*: Se observa que la variable se encuentra mayormente concentrada en 0 minutos, indicando que es algo atípico el despegar con atraso. Así mismo, se tiene que el promedio de atraso son 14 minutos, con una desviación estándar de 38 minutos; teniendo que el 50% de los pasajeros tuvo un vuelos sin atraso, y que el 75% tuvo un atraso de al menos 12 minutos, teniendo como máximo atraso 1.592 minutos.

In [30]:
fig = px.histogram(df_numerics_var, x = 'Departure Delay in Minutes', title = 'Departure Delay in Minutes Histogram')
fig.show()

* *Arrival Delay in Minutes*: Se observa que la variable se encuentra mayormente concentrada en 0 minutos, indicando que es algo atípico el aterrizar con atraso. Así mismo, se tiene que el promedio de atraso son 15 minutos, con una desviación estándar de 38 minutos; teniendo que el 50% de los pasajeros tuvo un vuelos sin atraso, y que el 75% tuvo un atraso de al menos 13 minutos, teniendo como máximo atraso 1.584 minutos. Esta variable se encuentra muy correlacionada con *Departure Delay in Minutes*, ya que lógicamente, el tiempo de atraso que se tiene al partir es similar al tiempo de atraso que tendrá al llegar.

In [31]:
fig = px.histogram(df_numerics_var, x = 'Arrival Delay in Minutes', title = 'Arrival Delay in Minutes Histogram')
fig.show()

**P3**

En vista de los datos y las distribuciones observadas, todas las variables deben ser transformadas. En particular, las que variables categoricas ordinales, como las correspondientes a niveles de satisfacción, y las continuas, como la edad, distancia de vuelo o retrasos tanto en inicio como llegada, deben ser escaladas, por ejemplo, por medio de *StandardScaler*, para que puedan ser comparables entre ellas; mientras que las variables categoricas nominales, como el tipo de viaje o la clase, deben ser transformadas por *One Hot*.

**P4**

Se observa alta correlación entre las variables:

* *Arrival Delay in Minutes* - *Departure Delay in Minutes* : 0.97
* *Ease of Online booking* - *Inflight wifi service* : 0.72
* *Inflight entertainment* - *Cleanliness* : 0.69
* *Cleanliness* - *Seat comfort* : 0.68
* *Food and drink* - *Cleanliness* : 0.66
* *Baggage handling* - *Inflight service* : 0.63
* *Inflight entertainment* - *Food and drink* : 0.62

In [10]:
df_numerics_var.corr()

id       Age  Flight Distance  \
id                                 1.000000  0.024276         0.095731   
Age                                0.024276  1.000000         0.101150   
Flight Distance                    0.095731  0.101150         1.000000   
Inflight wifi service             -0.021350  0.017572         0.007187   
Departure/Arrival time convenient -0.002577  0.039008        -0.019793   
Ease of Online booking             0.013705  0.024927         0.066585   
Gate location                     -0.001756 -0.001224         0.004086   
Food and drink                     0.002246  0.022470         0.057216   
Online boarding                    0.056320  0.207535         0.216355   
Seat comfort                       0.056135  0.158939         0.158620   
Inflight entertainment             0.005128  0.076808         0.129742   
On-board service                   0.055527  0.059467         0.109537   
Leg room service                   0.047001  0.042472         0.134848   
Baggage handling                   0.077104 -0.046668         0.064242   
Checkin service                    0.079611  0.034976         0.073001   
Inflight service                   0.079771 -0.048948         0.057966   
Cleanliness                        0.027865  0.052684         0.093521   
Departure Delay in Minutes        -0.019847 -0.009492         0.001517   
Arrival Delay in Minutes          -0.037604 -0.011283        -0.002838   

                                   Inflight wifi service  \
id                                             -0.021350   
Age                                             0.017572   
Flight Distance                                 0.007187   
Inflight wifi service                           1.000000   
Departure/Arrival time convenient               0.343414   
Ease of Online booking                          0.715727   
Gate location                                   0.335391   
Food and drink                                  0.134076   
Online boarding                                 0.456805   
Seat comfort                                    0.122673   
Inflight entertainment                          0.209171   
On-board service                                0.123733   
Leg room service                                0.160075   
Baggage handling                                0.121764   
Checkin service                                 0.043869   
Inflight service                                0.111062   
Cleanliness                                     0.131513   
Departure Delay in Minutes                     -0.016213   
Arrival Delay in Minutes                       -0.018254   

                                   Departure/Arrival time convenient  \
id                                                         -0.002577   
Age                                                         0.039008   
Flight Distance                                            -0.019793   
Inflight wifi service                                       0.343414   
Departure/Arrival time convenient                           1.000000   
Ease of Online booking                                      0.437551   
Gate location                                               0.445463   
Food and drink                                              0.004046   
Online boarding                                             0.069607   
Seat comfort                                                0.010475   
Inflight entertainment                                     -0.005432   
On-board service                                            0.068800   
Leg room service                                            0.011975   
Baggage handling                                            0.071662   
Checkin service                                             0.092419   
Inflight service                                            0.072624   
Cleanliness                                                 0.013094   
Departure Delay in Minutes                                  0.0

**P5**

Se escogieron las siguientes variables:

* *Age*: Se escogió la edad de los pasajeros porque no estaba correlacionada con el resto de las variables, lo que reduce la multicolinealidad y le da más estabilidad a los algoritmos de clustering que se quieran utilizar. 

* *Flight Distance*: Al igual que con la edad, se escogió esta variable porque no estaba correlacionada con el resto de ellas.

* *Online boarding*: Esta se encuentra correlacionada con otras variables, como por ejemplo *Inflight service* y *Ease of Online Booking*, entre otras; por lo que incorporaría de buena manera el resto de las características.

* *Seat comfort*: Al igual que *Online boarding*, esta variable se encuentra correlacionada con otras, como por ejempo *Food and drink* y *Inflight entertainment*, incorporando de buena manera estas variables, que no serán consideradas.

Las variables que se eliminaron fue porque no contaban con suficiente independencia, como las dos primeras; o porque estaban correlacionadas con una de estas dos últimas.

In [11]:
df_selected_var = df_numerics_var[['Age', 'Flight Distance', 'Online boarding', 'Seat comfort']]

## 3. Preprocesamiento 🎭

<center>
<img src="https://i.pinimg.com/originals/1e/a8/0e/1ea80e7cea0d429146580c7e91c5b944.gif" width=400>

Tras quedar satisfecho con los resultados presentados en el punto 2, el dueño de la empresa ha solicitado que se preprocesen los datos mediante un `pipeline`. Es crucial que este proceso tenga en cuenta las observaciones derivadas de los análisis anteriores. Adicionalmente, ha expresado su interés en visualizar el conjunto de datos en un gráfico de dos o tres dimensiones.

Basándose en los análisis realizados anteriormente:
1. Cree un `pipeline` que incluya PCA, utilizando las consideraciones mencionadas previamente para proyectar los datos a dos dimensiones. [4 puntos]
2. Grafique los resultados obtenidos y comente lo visualizado. [6 puntos]

**Respuestas:**

**P1**

Para poder crear un pipeline que incluya PCA, se debe aplicar una transformación previa a los datos, que corresponde a un escalamiento, utilizando StandardScaler.

In [12]:
from sklearn.decomposition import PCA
from sklearn.pipeline import Pipeline
from sklearn.preprocessing import StandardScaler
from sklearn.compose import ColumnTransformer

In [13]:
pipeline = Pipeline([
    ('scaler', StandardScaler()),
    ('pca', PCA(n_components=2))
])

In [14]:
df_transformed = pipeline.fit_transform(df_selected_var)

In [15]:
df_transformed

array([[ 0.39868591,  0.79900892],
       [ 1.91273605,  0.03011604],
       [-1.10375564,  0.65843408],
       ...,
       [ 0.26235398,  0.97919575],
       [ 2.59029791,  0.66302231],
       [ 2.17805482,  0.90884966]])

**P2**

In [17]:
fig = px.scatter(x = df_transformed[:, 0], 
                 y = df_transformed[:, 1],
                 labels={'x' : "Principal Component 1",
                         'y' : "Principal Component 2"},
                 title = 'Scatter Plot 2D PCA')
fig.show()

Dada la distribución de la varianza en el gráfico de PCA, donde se tiene que esta centrada en el espacio, puede sugerir que existe cierto grado de correlación entre las variables originales, donde tanto la primera y la segunda componente estan capturando la mayor cantidad de varianza en la data.

## 4. Outliers 🚫🙅‍♀️❌🙅‍♂️

<center>
<img src="https://joachim-gassen.github.io/images/ani_sim_bad_leverage.gif" width=250>

Con el objetivo de mantener la claridad en su análisis, Don Sergio le ha solicitado entrenar un modelo que identifique pasajeros con comportamientos altamente atípicos.

1. Utilice `IsolationForest` para clasificar las anomalías del dataset, configurando el modelo para que sólo el 1% de los datos sean considerados anómalos. Asegúrese de integrar esta tarea dentro de un `pipeline`. [3 puntos]

2. Visualice los resultados en el gráfico de dos dimensiones previamente creado. [3 puntos]

3. ¿Cómo evaluaría el rendimiento de su modelo en la detección de anomalías? [4 puntos]

**Respuestas:**

**P1**

In [18]:
from sklearn.ensemble import IsolationForest
isolation_forest_model = IsolationForest(contamination=0.01)
pipeline.steps.append(['isolation_forest', isolation_forest_model])

In [19]:
pipeline.fit(df_selected_var)

Pipeline(steps=[('scaler', StandardScaler()), ('pca', PCA(n_components=2)),
                ['isolation_forest', IsolationForest(contamination=0.01)]])

In [20]:
anomaly_data = pipeline.named_steps['isolation_forest'].predict(df_transformed)

In [21]:
df_transformed_nooutliers = pd.DataFrame(np.column_stack((df_transformed, anomaly_data)))
df_transformed_nooutliers = df_transformed_nooutliers[df_transformed_nooutliers[2] == 1]

**P2**

In [23]:
fig = px.scatter(x = df_transformed_nooutliers[0], 
                 y = df_transformed_nooutliers[1],
                 labels={'x' : "Principal Component 1",
                         'y' : "Principal Component 2"},
                 title = 'Scatter Plot 2D PCA')
fig.show()

**P3**

Para evaluar el rendimiento del modelo, con respecto a la detección de anomalías, utilizaría un modelo de clustering como puede ser K-means o GMM, con distintos hiperparámetros, con el fin de ir visualizando los ditintos grupos, antes y después de la eliminación de outliers. De esta manera, se puede utilizar AIC y BIC, para ver cuál de estos genera resultados más consistentes y eficientes.

## 5. Métricas de Desempeño 🚀

<center>
<img src="https://giffiles.alphacoders.com/219/219081.gif" width=300>

Motivado por incrementar su fortuna, Don Sergio le solicita entrenar un modelo que le permita segmentar a los pasajeros en grupos distintos, con el objetivo de optimizar las diversas campañas de marketing diseñadas por su equipo. Para ello, le se pide realizar las siguientes tareas:

1. Utilizar el modelo **Gaussian Mixture** y explore diferentes configuraciones de número de clústers, específicamente entre 3 y 8. Asegúrese de integrar esta operación dentro de un `pipeline`. [4 puntos]
2. Explique cuál sería el criterio adecuado para seleccionar el número óptimo de clústers. **Justifique de forma estadistica y a traves de gráficos.** [6 puntos]

> **HINT:** Se recomienda investigar sobre los criterios AIC y BIC para esta tarea.

**Respuestas**

**P1**

In [24]:
labels_fited = []
for n_component in range(3,9):
    gmm = GaussianMixture(n_components = n_component, random_state=666)
    gmm.fit(df_transformed_nooutliers)
    gmm_labels = gmm.predict(df_transformed_nooutliers)
    labels_fited.append(gmm_labels)
    print('N components:', {n_component})
    print('aic: ', gmm.aic(df_transformed_nooutliers))
    print('bic: ', gmm.bic(df_transformed_nooutliers))

N components: {3}
aic:  -1429832.4521343845
bic:  -1429531.9536702898
N components: {4}
aic:  -1432072.9201206043
bic:  -1431668.801496477
N components: {5}
aic:  -1435995.9634483894
bic:  -1435488.2246642294
N components: {6}
aic:  -1436606.506676799
bic:  -1435995.1477326064
N components: {7}
aic:  -1437253.5567139848
bic:  -1436538.5776097595
N components: {8}
aic:  -1438239.4788380363
bic:  -1437420.8795737785


**P2**

Para la selección optima, se visualizaron los gráficos correspondientes a los clusters generados, considerando entre 3 y 8 componentes, observando en cuales estas se diferenciaban más entre ellos. Además, se calcula el AIC y el BIC de cada modelo. De esta manera, a través del análisis de los gráficos y los valores de AIC, que indica que tan bien se ajusta el modelo a los datos, penalizando la complejidad del modelo; y BIC, que penaliza de igual manera que el AIC, pero no tan fuertemente; se puede determinar un número optimo de componentes a escoger.

In [54]:
labels_fited = []
for n_component in range(3,9):
    gmm = GaussianMixture(n_components = n_component, random_state=666)
    gmm.fit(df_transformed_nooutliers)
    gmm_labels = gmm.predict(df_transformed_nooutliers)
    labels_fited.append(gmm_labels)
    print('N components:', {n_component})
    print('aic: ', gmm.aic(df_transformed_nooutliers))
    print('bic: ', gmm.bic(df_transformed_nooutliers))

N components: {3}
aic:  -1430020.827778822
bic:  -1429720.328818479
N components: {4}
aic:  -1433795.1649126494
bic:  -1433391.045621154
N components: {5}
aic:  -1436386.8409507058
bic:  -1435879.1013280575
N components: {6}
aic:  -1436072.012071698
bic:  -1435460.652117897
N components: {7}
aic:  -1437156.7755981546
bic:  -1436441.795313201
N components: {8}
aic:  -1438568.411220782
bic:  -1437749.8106046757


In [ ]:
for i in range(len(labels_fited)):
    fig = px.scatter(x = df_transformed_nooutliers[0], 
                     y = df_transformed_nooutliers[1],
                     labels={'x' : "Principal Component 1",
                             'y' : "Principal Component 2"},
                     color = labels_fited[i],
                     title = 'Scatter Plot 2D PCA')
    fig.show()

## 6. Análisis de resultados 📊

<center>
<img src="https://i.gifer.com/7wTk.gif" width=300>

Una vez identificado el número óptimo de clústers, se le pide realizar lo siguiente:

1. Utilizar la proyección en dos dimensiones para visualizar cada clúster claramente. [2 puntos]

2. ¿Es posible distinguir claramente entre los clústers generados? [2 puntos]

3. Proporcionar una descripción breve de cada clúster utilizando estadísticas descriptivas básicas, como la media y la desviación estándar, para resumir las características de las variables utilizadas en estos algoritmos. [2 puntos]

4. Proceda a visualizar los clústers en tres dimensiones para una perspectiva más detallada. [2 puntos]

5. ¿Cómo afecta esto a sus conclusiones anteriores? [2 puntos]

**Respuestas:**

**P1**

Se escoge 5 como número óptimo de componentes.

In [26]:
fig = px.scatter(x = df_transformed_nooutliers[0], 
                 y = df_transformed_nooutliers[1],
                 labels={'x' : "Principal Component 1",
                         'y' : "Principal Component 2"},
                 color = labels_fited[2],
                 title = 'Scatter Plot 2D PCA')
fig.show()

**P2**

Es posible distinguir los clústers generados. Se escogió esta cantidad, ya que después se genera un clustar al interior de los datos, provocando que puedan existir dificultades en la clasificación al colindar con todas las demás clases.

**P3**

A continuación se entrega una descripción de los clusters detectados:

* Cluster 0: Este clúster tiene una media relativamente alta en la primera variable (0.654) y una media baja en la segunda variable (-0.928). Además, la variabilidad en ambas variables es moderada, con desviaciones estándar de aproximadamente 0.51 y 0.49 respectivamente.

* Cluster 1: Aquí se observa una media más baja en la primera variable (0.228) en comparación con los otros clústeres, mientras que la media en la segunda variable es relativamente alta (0.638). La variabilidad en este clúster también es moderada, con desviaciones estándar de alrededor de 0.46 y 0.55 respectivamente.

* Cluster 2: Este clúster muestra medias bastante bajas en ambas variables (-0.936 y -0.807 respectivamente). La variabilidad en este clúster es similar a la del clúster 1, con desviaciones estándar de aproximadamente 0.46 y 0.56 para las dos variables.

* Cluster 3: Tiene una media muy baja en la primera variable (-1.613) y una media más alta en la segunda variable (0.885). La variabilidad en este clúster es algo más alta que en los clústeres anteriores, con desviaciones estándar de alrededor de 0.55 y 0.63 respectivamente.

* Cluster 4: Este clúster tiene las medias más altas en ambas variables (1.739 y 0.329 respectivamente). La variabilidad en este clúster es similar a la del clúster 0, con desviaciones estándar de aproximadamente 0.52 y 0.54 respectivamente.

En resumen, cada clúster se caracteriza por diferentes niveles de medias y variabilidades en las dos variables utilizadas en el algoritmo de clustering.

In [27]:
df_transformed_nooutliers['labels'] = labels_fited[2]

In [28]:
labels_sorted = df_transformed_nooutliers['labels'].unique()
labels_sorted.sort()

for i in labels_sorted:
    df_subset = df_transformed_nooutliers[df_transformed_nooutliers['labels'] == i]
    
    print(f'Media de cluster {i}: ')
    print(df_subset[[0, 1]].mean()[0])
    print(df_subset[[0, 1]].mean()[1])
    print()
    print(f'Std de cluster {i}: ')
    print(df_subset[[0, 1]].std()[0])
    print(df_subset[[0, 1]].std()[1])
    print()

Media de cluster 0: 
-0.9243486957471282
-0.8162486052902673

Std de cluster 0: 
0.46575152986691315
0.5601100470707029

Media de cluster 1: 
0.22023654706930643
0.6320303652294883

Std de cluster 1: 
0.4500598515009856
0.5562138322934128

Media de cluster 2: 
1.7342839164335326
0.34049294695979293

Std de cluster 2: 
0.5093669391908348
0.5640869474569068

Media de cluster 3: 
-1.6114603847144766
0.8617489786997411

Std de cluster 3: 
0.5516402122265771
0.6103715387059558

Media de cluster 4: 
0.6633629860454195
-0.9347599198458282

Std de cluster 4: 
0.4956905350275535
0.49644976554408976



**P4**

Para esto, se redefine el pca, con 3 componentes y se integran denuevo los pasos:

In [29]:
pipeline = Pipeline([
    ('scaler', StandardScaler()),
    ('pca', PCA(n_components=3))
])

In [30]:
df_transformed = pipeline.fit_transform(df_selected_var)

In [31]:
isolation_forest_model = IsolationForest(contamination=0.01)
pipeline.steps.append(['isolation_forest', isolation_forest_model])

In [32]:
pipeline.fit(df_transformed)

Pipeline(steps=[('scaler', StandardScaler()), ('pca', PCA(n_components=3)),
                ['isolation_forest', IsolationForest(contamination=0.01)]])

In [33]:
anomaly_data = pipeline.named_steps['isolation_forest'].predict(df_transformed)

In [34]:
df_transformed_nooutliers = pd.DataFrame(np.column_stack((df_transformed, anomaly_data)))
df_transformed_nooutliers = df_transformed_nooutliers[df_transformed_nooutliers[3] == 1]

In [35]:
gmm = GaussianMixture(n_components = 5, random_state=666)
gmm.fit(df_transformed_nooutliers)
gmm_labels = gmm.predict(df_transformed_nooutliers)

In [37]:
fig = px.scatter_3d(x = df_transformed_nooutliers[0], 
                    y = df_transformed_nooutliers[1],
                    z = df_transformed_nooutliers[2],
                    labels={'x' : "Principal Component 1",
                            'y' : "Principal Component 2",
                            'z' : "Principal Component 3"},
                    color = gmm_labels,title = 'Scatter Plot 3D PCA')
fig.show()

**P5**

Al observar el gráfico en 3D generado, se observa un cluster que se colinda y se mezcla con otro (cluster 0 con el 3). De esta manera, puede que sea un mejor número de cluster 4, haciendo que este cluster conflictivo pase a formar parte de los otros más cercanos.